### PR TITLE
Better Statements

### PR DESCRIPTION
**Pull request by corymcdonald**
#### Features

Improve the statements for publishers. This is most likely going to be closed and re-open when it's finalized

### CODE REVIEW

File: app/javascript/packs/StatementDetails.jsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import StatementDetails from "../views/statements/statements/StatementDetails";
+
+document.addEventListener("DOMContentLoaded", () => {
+  const element = document.querySelector("#statement_details");
+  console.log(element.dataset);
```
Comment (yachtcaptain23) [inline app/javascript/packs/StatementDetails.jsx]: Debug statement

File: app/controllers/publishers/statements_controller.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      else
+        @statement_period = publisher_statement_period(@transactions)
+        statement_file_name = publishers_statement_file_name(@statement_period)
+        statement_string = render_to_string :layout => "statement"
+        send_data statement_string, filename: statement_file_name, type: "application/html"
+      end
+    end
+
+
+    def publisher
```
Comment (yachtcaptain23) [inline app/controllers/publishers/statements_controller.rb]: Might be worth `def protect` to application_controller and adding `before_action :protect`

Comment (corymcdonald) [reply app/controllers/publishers/statements_controller.rb]: There's a before_action that ensures a user is currently logged in

Does that satisfy the intended purpose of the `protect` method?

File: app/javascript/views/statements/statements/StatementDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+          {this.props.statement.details.map((detail, index) => (
+            <StatemenDetail detail={detail} index={index} />
+          ))}
+        </div>
+      </Details>
+    );
+  }
+}
+
+const StatemenDetail = props => (
```
Comment (yachtcaptain23) [inline app/javascript/views/statements/statements/StatementDetails.tsx]: Misspelling

File: app/models/views/user/statements.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      end
+
+      def as_json(*)
+        {
+          overviews: overviews,
+        }
+      end
+
+      def build_statements(publisher, details_date)
+        @overviews = []
```
Comment (yachtcaptain23) [inline app/models/views/user/statements.rb]: I think you can keep this as a local variable

File: app/javascript/views/statements/statements/StatementDetails.tsx (hunk 3 above)
Comment (corymcdonald) [reply app/javascript/views/statements/statements/StatementDetails.tsx]: https://github.com/brave-intl/publishers/pull/2209/commits/1b300b11587f843bc178734ac0de4a8492c8e14e

**Comment by corymcdonald** [comment]
Discussed offline, going to develop a process for localizing rest of application

File: app/controllers/publishers/statements_controller.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    end
+
+    private
+
+    def publisher
+      return Publisher.find(params[:id]) if current_publisher.admin?
+      current_publisher
+    end
+
+    def protect
```
Comment (yachtcaptain23) [inline app/controllers/publishers/statements_controller.rb]: I think we should move `def protect` from PublisherController to ApplicationController

File: app/models/views/user/statements.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        # Then we can group the payouts into monthly periods for the statement
+        periods = payouts.group_by { |x| x.created_at.at_beginning_of_month }
+
+        periods.each do |payment_month, payout_entries|
+          # We find the previous month that was paid, or the first statement that was generated
+          period_start = periods.keys.reverse.detect { |k| k < payment_month } || @statements.first.created_at.at_beginning_of_month
+
+          # Sum all the BAT amounts
+          total_amount = payout_entries.sum { |x| x.amount.abs }
```
Comment (yachtcaptain23) [inline app/models/views/user/statements.rb]: I think it should be `payout_entries.sum { |x| x.amount > 0 ? x.amount : 0 }`

File: app/services/publisher_statement_getter.rb [hunk 7]
```diff
@@ -1,53 +1,100 @@
...
+
+    # Replicate the similar behavior of Eyeshade. Aggregates all the contribution amounts for the month, and puts them into one entry
+    uphold.group_by { |u| u.created_at.at_end_of_month }.each do |date, entries|
+      entries.group_by { |e| e.channel }.each do |channel, channel_entries|
+        amount = channel_entries.sum { |x| x.amount }
+
+        uphold << Statement.new(
+          channel: channel,
+          transaction_type: Statement::UPHOLD_CONTRIBUTION_SETTLEMENT,
+          amount: -amount,
```
Comment (yachtcaptain23) [inline app/services/publisher_statement_getter.rb]: Why is this negative? (Preferable you add a comment in-line to explain)

File: app/services/publisher_statement_getter.rb [hunk 8]
```diff
@@ -1,53 +1,100 @@
...
+          amount: transaction.origin.dig("amount")&.to_d,
+          settlement_currency: transaction.destination.dig("currency"),
+          settlement_amount: transaction.destination.dig("amount")&.to_d,
+          created_at: transaction.createdAt.to_date,
+        )
+      end
+    end
+
+    # Replicate the similar behavior of Eyeshade. Aggregates all the contribution amounts for the month, and puts them into one entry
+    uphold.group_by { |u| u.created_at.at_end_of_month }.each do |date, entries|
```
Comment (yachtcaptain23) [inline app/services/publisher_statement_getter.rb]: I'm very confused by this block. So we're going through a list of uphold contributions and somehow the settlement statements gets appended in this block?

**Comment by yachtcaptain23** [comment]
Where does `publishers/statements/statement.html.slim` get called to be rendered?

File: app/services/publisher_statement_getter.rb (hunk 8 above)
Comment (corymcdonald) [reply app/services/publisher_statement_getter.rb]: When we mark a transaction as paid out then we subtract from the Publishers current balance on Eyeshade so the transactions appear as negative. 

In order to actually get a valid grouping for periods for which months we paid out I select all the transactions that are negative, then group them by then select the transactions which are within the start and end of that grouping.

This was the easiest way for finding the ranges for when publishers got paid. So for Uphold it’s a little different. Technically we are have technically already paid these transactions out as they happen in real time. However if we did mark them as negative BAT amounts and the positive amount that the user got paid then our monthly groupings would be wrong. 

In order to keep the same groupings and also include the summary for how many Uphold tips there were I summarize them after I append them to all the transactions.

File: app/controllers/publishers/statements_controller.rb (hunk 5 above)
Comment (corymcdonald) [reply app/controllers/publishers/statements_controller.rb]: https://github.com/brave-intl/publishers/pull/2209/commits/152975f8a26ad1f7223146f0a0b7338b3bdfd1de

File: app/models/views/user/statements.rb (hunk 6 above)
Comment (corymcdonald) [reply app/models/views/user/statements.rb]: I don't think so, basically Eyeshade returns negative amounts (and a positive conversion with currency) when we pay users out, so this sums up all the negative amounts that we paid out.

**Comment by corymcdonald** [comment]
>Where does publishers/statements/statement.html.slim get called to be rendered?

Good catch :) It doesn't get rendered anymore.

File: app/services/publisher_statement_getter.rb (hunk 7 above)
Comment (corymcdonald) [reply app/services/publisher_statement_getter.rb]: https://github.com/brave-intl/publishers/pull/2209/commits/35058de3041adc5cf544b4d3bce3a4d4594a66da

File: app/models/views/user/statements.rb (hunk 4 above)
Comment (corymcdonald) [reply app/models/views/user/statements.rb]: I don't think so, because this is what is being serialized and sent as json to the view

**Pull request merged**